src/wasm-v8.cc

```diff
@@ -642,16 +597,16 @@ struct FuncTypeImpl : ExternTypeImpl {
   }
 };
 
-template<> struct implement<FuncType> { using type = FuncTypeImpl; };
+static_assert(std::is_standard_layout<ExternTypeImpl<FuncTypeImpl, FuncType, ExternKind::FUNC>>::value);
```

Comment:
Why are these assertions needed? Can they be avoided somehow?

Reply:
Of course asserts are just that, their only purpose is to sometimes make the build fail, so they can be removed. But in this case they're indicating something very important.

Recall my original plan was `Config::method()` can safely cast `Config*` to `ConfigImpl*` because the only way to get a `Config*` is through `Config::make()` which always actually creates a `ConfigImpl`?

`ExternType` is the base of a type hierarchy. In C++ when you create an object, it starts by creating the base object first, so that's a second way to create `ExternType`. This means that when you see an `ExternType*` you don't know whether it's an `ExternTypeImpl*` or a casted `FuncType*`. Oops. How do you implement ExternType::methods()?

The obvious fix of saying "well, `FuncTypImpl` derives from `ExternTypeImpl` too" doesn't work either because now you have two distinct copies of `ExternType`, and a given `ExternType*` might be the one that's a base of `FuncType` of `FuncTypeImpl` or the one that's a base of `ExternTypeImpl` of `FuncTypeImpl`. Two copies of a base objects usually have to have distinct addresses (C++20 adds an attribute for this but it's a suggestion).

Quick aside, there's a common C++ fix for this: virtual inheritance. When inheriting virtually, you merge two copies of a base into one copy. Tada! But I'm continuing to assume we can't use virtual.

So here's what I'm going to do: we add a simple `struct ExternTypeKind` that allows us to tell what the most-derived type of the object is. Once we know that, we're home free because we know what to cast to. Now, we need to be able to go from `ExternType*` or `FuncType*` etc. to the `ExternTypeKind*` even though we don't know what casts have been performed previously. The standard term for this is _pointer-interconvertible_ and it will allow us to use `reinterpret_cast` to correctly cast between them. `ExternTypeImpl` exists to follow those rules, one of which is being a standard-layout type. That's what the static assert is checking.

Then we can derive `FuncTypeImpl` on top, and that can do whatever it likes including things which cause static_cast<> between `FuncTypeImpl` and `ExternType` to have a pointer adjustment.

Reference: https://eel.is/c++draft/basic.compound#4.3 . Also since it doesn't link to "standard-layout class": https://eel.is/c++draft/class.prop#3